libexec/ch.py

```diff
@@ -0,0 +1,247 @@
+            tf_info = v.getmembers()
+            tf_members = list()
+            for m in tf_info:
+                if m.isdev():
+                    dev_ct += 1
+                    DEBUG('ignoring device file {}'.format(m.name))
+                elif re.search('\.wh\..*', m.name):
+                    wh_ct += 1
+                    DEBUG('whiteout found: {}'.format(m.name))
+                    wh = os.path.basename(m.name)
+                    wh_dirname = os.path.dirname(m.name)
+                    wh_target = os.path.join(wh_dirname, wh.split('.wh.')[-1])
+                    if os.path.exists(wh_target):
+                        DEBUG('removing {}'.format(wh_target))
+                        if os.path.isdir(wh_target):
+                            shutil.rmtree(wh_target)
+                        else:
+                           os.remove(wh_target)
+                    else:
+                       FATAL("whiteout target {} doesn't exist".format(wh_target))
+                elif re.search('^\.\./.*', m.name) or re.search('^/.*', m.name):
+                    FATAL("dangerous extraction path '{}'".format(m.name))
+                else:
+                    tf_members.append(m)
+
+            INFO('extracting layer {}'.format(k))
+            v.extractall(members=tf_members)
+
+        print('image successfully unpacked.')
+        if dev_ct > 0:
+            INFO("{} device files ignored.".format(dev_ct))
+        if wh_ct > 0:
+            INFO("{} whiteout files handled.".format(wh_ct))
+
+
+class MySession:
+    def __init__(self, image):
+        self.token   = self.get_token(image)
+        self.headers = self.get_headers()
+
+    def get_headers(self):
+        return {'Authorization': 'Bearer {}'.format(self.token)}
+
+    def get_token(self, image):
+        tokenService = '{}/token?service={}'.format(authBase, authService)
+        scopeRepo    = '&scope=repository:{}:pull'.format(image.name)
+        authURL      = tokenService + scopeRepo
+        return session.get(authURL, proxies=PROXIES).json()['token']
+
+
+## Supporting functins ##
+
+def DEBUG(*args, **kwargs):
+    color("36m", sys.stderr)
+    print(flush=True, file=sys.stderr, *args, **kwargs)
+    color_reset(sys.stderr)
+
+def ERROR(*args, **kwargs):
+    color("31m", sys.stderr)
+    print(flush=True, file=sys.stderr, *args, **kwargs)
+    color_reset(sys.stderr)
+
+def FATAL(*args, **kwargs):
+    ERROR(*args, **kwargs)
+    sys.exit(1)
+
+def INFO(*args, **kwargs):
+        print(flush=True, *args, **kwargs)
+
+def color(color, fp):
+    if (fp.isatty()):
+        print("\033[" + color, end="", flush=True, file=fp)
+
+def color_reset(*fps):
+    for fp in fps:
+        color("0m", fp)
+
+def log():
+    logging.basicConfig(format='%(levelname)s:%(message)s')
+    HTTPConnection.debuglevel = 1
+    logging.getLogger().setLevel(logging.DEBUG)
+    rlog = logging.getLogger("requests.packages.urllib3").setLevel(logging.DEBUG)
```

Comment:
Not sure how (if) the HTTPConnection debug output can be colored. Seems like a pain.

Reply:
I would not worry about that for the time being.